src/native/eventpipe/ep.c

```diff
@@ -1313,15 +1317,25 @@ ep_delete_provider (EventPipeProvider *provider)
 	// Take the lock to make sure that we don't have a race
 	// between disabling tracing and deleting a provider
 	// where we hold a provider after tracing has been disabled.
+	bool wait_for_provider_callbacks_completion = false;
 	EP_LOCK_ENTER (section1)
 		if (enabled ()) {
 			// Save the provider until the end of the tracing session.
 			ep_provider_set_delete_deferred (provider, true);
+
+			// the callback func must be previously set to null,
+			// otherwise callbacks might never stop coming
+			EP_ASSERT (provider->callback_func == NULL);
+			if (provider->uninvoked_prepared_callbacks_counter > 0)
+				wait_for_provider_callbacks_completion = true;
 		} else {
 			config_delete_provider (ep_config_get (), provider);
 		}
 	EP_LOCK_EXIT (section1)
 
+	if (wait_for_provider_callbacks_completion)
```

Comment:
It's not immediately obvious to me why we need to wait here, can you add a comment explaining what it prevents?

Reply:
From looking at the issue I see that the issue is that the GCHandle gets deleted on the managed code side of things, but how does this blocking wait prevent that?

Reply:
Right before managed code deletes the GCHandle it calls DeleteProvider() which comes into this code. Blocking ensures that all the callbacks running asynchronously which might have tried to use that GCHandle are complete before the GCHandle gets deleted.

Separately it also just gives ETW and EventPipe the same blocking behavior for Unregister() so that users+ourselves don't have to reason about two different synchronization patterns depending on whether EventPipe or ETW was used.

Reply:
Added a comment there describing the blocking behavior